velox/adapters/s3fs/S3FileSystem.cpp

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/adapters/s3fs/S3FileSystem.h"
+#include "velox/common/file/File.h"
+#include "velox/core/Context.h"
+
+#include <fmt/format.h>
+#include <glog/logging.h>
+#include <memory>
+#include <mutex>
+#include <stdexcept>
+
+#include <aws/core/Aws.h>
+#include <aws/core/Region.h>
+#include <aws/core/auth/AWSCredentialsProviderChain.h>
+#include <aws/core/http/HttpResponse.h>
+#include <aws/core/utils/logging/ConsoleLogSystem.h>
+#include <aws/s3/S3Client.h>
+#include <aws/s3/model/GetObjectRequest.h>
+
+namespace facebook::velox {
+
+constexpr std::string_view kSep{"/"};
+constexpr std::string_view s3Scheme{"s3:"};
+
+bool inline isS3File(const std::string_view filename) {
+  return (filename.substr(0, s3Scheme.size()) == s3Scheme);
+}
+
+inline Aws::String getAwsString(const std::string& s) {
+  return Aws::String(s.begin(), s.end());
+}
+
+inline std::string_view fromAwsString(const Aws::String& s) {
+  return {s.data(), s.length()};
+}
+
+// Implement the S3ReadFile
+class S3ReadFile final : public ReadFile {
+ public:
+  S3ReadFile(std::string path, std::shared_ptr<Aws::S3::S3Client> client)
+      : path_(std::move(path)), client_(client) {}
+
+  std::string_view pread(uint64_t offset, uint64_t length, Arena* arena)
+      const override {
+    char* pos = arena->reserve(length);
+    preadInternal(offset, length, pos);
+    return {pos, length};
+  }
+
+  std::string_view pread(uint64_t offset, uint64_t length, void* buf)
+      const override {
+    preadInternal(offset, length, static_cast<char*>(buf));
+    return {static_cast<char*>(buf), length};
+  }
+
+  std::string pread(uint64_t offset, uint64_t length) const override {
+    // TODO: use allocator that doesn't initialize memory?
+    std::string result(length, 0);
+    char* pos = result.data();
+    preadInternal(offset, length, pos);
+    return result;
+  }
+
+  uint64_t preadv(
+      uint64_t offset,
+      const std::vector<folly::Range<char*>>& buffers) override {
+    throw std::runtime_error("failure in S3ReadFile::preadv.");
+    return -1;
+  }
+
+  uint64_t size() const override {
+    throw std::runtime_error("failure in S3ReadFile::size.");
```

Comment:
S3 has no API to return a file's size, or it's just not implemented yet? For the latter, we usually use `VELOX_NYI()` macro.

Reply:
It's just not implemented yet. I will implement it in this PR. Thanks for `VELOX_NYI()`. I will use that at other places.